android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboardDownloaderActivity.java

```diff
@@ -0,0 +1,446 @@
+          int result = 0;
+          for (String url : urls) {
+            String directory = KMDefault_AssetPackages + File.separator + pkgID;
+            File dir = new File(directory);
+            String filename = "";
+            if (url.endsWith(".js")) {
+
+              int start = kbFilename.lastIndexOf("/");
+              if (start < 0) {
+                start = 0;
+              } else {
+                start++;
+              }
+              if (!kbFilename.contains("-")) {
+                filename = kbFilename.substring(start, kbFilename.length() - 3) + "-" + kbVersion + ".js";
+              } else {
+                filename = kbFilename.substring(start);
+              }
+            }
+
+            result = FileDownloader.download(context, url, directory, filename);
+            if (result < 0) {
+              ret = -1;
+              break;
+            }
+          }
+        } catch (Exception e) {
+          ret = -1;
+          Log.e("Keyboard download", "Error: " + e);
+        }
+
+        return ret;
+      }
+
+      @Override
+      protected void onProgressUpdate(Integer... progress) {
+        // Do nothing
+      }
+
+      @Override
+      protected void onPostExecute(Integer result) {
+        if (showProgressDialog) {
+          if (progressDialog != null && progressDialog.isShowing()) {
+            try {
+              progressDialog.dismiss();
+              progressDialog = null;
+            } catch (Exception e) {
+              progressDialog = null;
+            }
+          }
+        }
+
+        notifyListeners(KeyboardEventHandler.EventType.KEYBOARD_DOWNLOAD_FINISHED, result);
+        super.onPostExecute(result);
+      }
+
+      /**
+       * Notify listeners when an event happens
+       * @param eventType
+       * @param result
+       */
+      protected void notifyListeners(KeyboardEventHandler.EventType eventType, int result) {
+        if (kbDownloadEventListeners != null) {
+          HashMap<String, String> keyboardInfo = new HashMap<String, String>();
+          keyboardInfo.put(KMManager.KMKey_PackageID, pkgID);
+          keyboardInfo.put(KMManager.KMKey_KeyboardID, kbID);
+          keyboardInfo.put(KMManager.KMKey_LanguageID, langID);
+          keyboardInfo.put(KMManager.KMKey_KeyboardName, kbName);
+          keyboardInfo.put(KMManager.KMKey_LanguageName, langName);
+          keyboardInfo.put(KMManager.KMKey_KeyboardVersion, kbVersion);
+          keyboardInfo.put(KMManager.KMKey_CustomKeyboard, kbIsCustom);
+          keyboardInfo.put(KMManager.KMKey_Font, font);
+          if (oskFont != null)
+            keyboardInfo.put(KMManager.KMKey_OskFont, oskFont);
+          KeyboardEventHandler.notifyListeners(kbDownloadEventListeners, eventType, keyboardInfo, result);
+        }
+      }
+    }.execute();
+  }
+
+  public static boolean isCustom(String u) {
+    boolean ret = false;
+    if (u != null && !u.contains(KMKeyboardDownloaderActivity.kKeymanApiBaseURL) &&
```

Comment:
Could always just return the boolean value inside the `if()` instead... though I suppose it is a long expression.

Reply:
@mcdurdin Is this the proper criteria in determining a URL goes with a Keyman server vs custom ad-hoc?

Reply:
I guess? Why do you need to differentiate?

Reply:
One of the keyboard info keys tracks if a keyboard is custom

Reply:
I'd prefer that distinction to be stored from an earlier base point (i.e. when the keyboard catalogue is constructed) rather than heuristically based on the URL.